hoomd/hpmc/ShapeMoves.h

```diff
@@ -0,0 +1,870 @@
+                    std::shared_ptr<IntegratorHPMCMono<Shape>> mc)
+        : ShapeMoveBase<Shape>(sysdef, mc), m_num_params(0)
+        {
+        m_params.resize(this->m_ntypes);
+        m_params_backup.resize(this->m_ntypes);
+        this->m_det_inertia_tensor = 1.0;
+        }
+
+    void prepare(uint64_t timestep)
+        {
+        m_params_backup = m_params;
+        }
+
+    void update_shape(uint64_t timestep,
+                      Scalar& stepsize,
+                      const unsigned int& type_id,
+                      typename Shape::param_type& shape,
+                      hoomd::RandomGenerator& rng)
+        {
+        for (unsigned int i = 0; i < m_params[type_id].size(); i++)
+            {
+            Scalar a = fmax(-stepsize, -(m_params[type_id][i]));
+            Scalar b = fmin(stepsize, (1.0 - m_params[type_id][i]));
+            hoomd::UniformDistribution<Scalar> uniform(a, b);
+            Scalar r = hoomd::detail::generate_canonical<double>(rng);
+            Scalar x = (r < this->m_move_probability) ? uniform(rng) : 0.0;
+            m_params[type_id][i] += x;
+            }
+        pybind11::object d = m_python_callback(type_id, m_params[type_id]);
+        pybind11::dict shape_dict = pybind11::cast<pybind11::dict>(d);
+        shape = typename Shape::param_type(shape_dict);
+        detail::MassProperties<Shape> mp(shape);
+        this->m_det_inertia_tensor = mp.getDetInertiaTensor();
+        }
+
+    void retreat(uint64_t timestep)
+        {
+        // move has been rejected.
+        std::swap(m_params, m_params_backup);
+        }
+
+    pybind11::list getParams(std::string typ)
+        {
+        unsigned int type_id = this->getValidateType(typ);
+        pybind11::list ret;
+        for (unsigned int i = 0; i < m_params[type_id].size(); i++)
+            {
+            ret.append(m_params[type_id][i]);
+            }
+        return ret;
+        }
+
+    void setParams(std::string typ, pybind11::list params)
+        {
+        unsigned int type_id = this->getValidateType(typ);
+        auto N = pybind11::len(params);
+        m_params[type_id].resize(N);
+        m_params_backup[type_id].resize(N);
+        for (unsigned int i = 0; i < N; i++)
+            {
+            m_params[type_id][i] = params[i].cast<Scalar>();
+            m_params_backup[type_id][i] = params[i].cast<Scalar>();
+            }
+        }
+
+    pybind11::dict getTypeParams()
+        {
+        pybind11::dict ret;
+        for (unsigned int type_id = 0; type_id < this->m_ntypes; type_id++)
+            {
+            std::string type_name = this->m_sysdef->getParticleData()->getNameByType(type_id);
+            pybind11::list l;
+            for (unsigned int i = 0; i < m_params[type_id].size(); i++)
+                {
+                l.append(m_params[type_id][i]);
+                }
+            ret[type_name.c_str()] = l;
+            }
+        return ret;
+        }
```

Comment:
Why is this useful? Rather than just querying the type parameter in Python?

Reply:
see https://github.com/glotzerlab/hoomd-blue/pull/989#discussion_r846688210

Reply:
removed as per https://github.com/glotzerlab/hoomd-blue/pull/989#discussion_r849351912